src/test/ui/async-await/async-fn-nonsend.rs

```diff
@@ -36,7 +36,7 @@ async fn non_send_temporary_in_match() {
 }
 
 async fn non_sync_with_method_call() {
-    // FIXME: it'd be nice for this to work.
+    // FIXME: it would be nice for this to work
```

Comment:
Yes, currently we do consider everything to be dropped in unreachable code, except for the start node where we consider everything to not be dropped.

I think the main thing missing is to handle functions that return `!` better.

I think it makes sense to do this as a follow up PR to this one though. What do you think?

Reply:
I tried replacing `panic!()` with `loop{}` and noticed I had a bug in my handling of empty loops. After adding a workaround for that issue, then the version of this test with `loop{}` works. So, I think the only missing thing is handling `!`.

Reply:
It turns out handling `!` return types wasn't actually that hard, so I pushed a couple more patches that do that.

However, the thing this test case is testing isn't unreachable code, but the borrow introduced by the method call in the `if` test. I rewrote the test in a way that still exposes this behavior.

I think the reason this doesn't work currently is because the MIR generation is super conservative about borrows and basically assumes anything that has ever been borrowed might be live across a yield point. So, I think it could be made to work, but it's probably outside the scope of drop tracking.

Reply:
That sounds right